util/src/main/java/org/datacommons/util/StringUtil.java

```diff
@@ -25,25 +25,34 @@
 import java.util.ArrayList;
 import java.util.List;
 import java.util.Locale;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Set;
 import java.util.regex.Pattern;
 import org.apache.commons.text.StringEscapeUtils;
 
 // Common set of utils to handle strings
 public class StringUtil {
   // From https://docs.oracle.com/javase/8/docs/api/java/time/format/DateTimeFormatterBuilder.html
-  private static final List<String> DATE_PATTERNS =
+  // Pattern specification:
+  // https://docs.oracle.com/javase/8/docs/api/java/time/format/DateTimeFormatterBuilder.html#appendPattern-java.lang.String-
+  // Key is the pattern, value is a list of lengths that this pattern could
+  // potentially match. This is used in filtering out impossible date checks by
+  // comparing the length of the input date string against the Set<Integer> here.
+  private static final List<Entry<String, Set<Integer>>> DATE_PATTERNS =
       List.of(
-          "yyyy",
-          "yyyy-MM",
-          "yyyyMM",
-          "yyyy-M",
-          "yyyy-MM-dd",
-          "yyyyMMdd",
-          "yyyy-M-d",
-          "yyyy-MM-dd'T'HH:mm",
-          "yyyy-MM-dd'T'HH:mm:ss",
-          "yyyy-MM-dd'T'HH:mm:ss.SSS",
-          "yyyy-MM-dd'T'HH:mm:ss.SSSXXX");
+          // Sort these entries starting with the most common cases for runtime
+          // speed performance.
+          Map.entry("yyyy-M", Set.of(6, 7)),
+          Map.entry("yyyy-M-d", Set.of(8, 9, 10)),
+          Map.entry("yyyy", Set.of(4)),
+          Map.entry("yyyyMM", Set.of(6)),
+          Map.entry("yyyyMMdd", Set.of(8)),
```

Comment:
nit: Order this as:
- yyyy
- yyyy-M
- yyyy-M-d
- yyyyMM
- yyyymmdd

So the 3 most common formats come on top.

Reply:
done.